src/processor/operator/arrow_result_collector.cpp

```diff
@@ -0,0 +1,102 @@
+#include "processor/operator/arrow_result_collector.h"
+
+#include "common/arrow/arrow_row_batch.h"
+#include "main/query_result/arrow_query_result.h"
+
+using namespace kuzu::common;
+
+namespace kuzu {
+namespace processor {
+
+bool ArrowResultCollectorLocalState::advance() {
+    for (int64_t i = (int64_t)chunks.size() - 1; i >= 0; --i) {
+        chunkCursors[i]++;
+        if (chunkCursors[i] < chunks[i]->state->getSelSize()) {
+            return true;
+        }
+        chunkCursors[i] = 0;
+    }
+    return false;
+}
+
+void ArrowResultCollectorLocalState::fillTuple() {
+    KU_ASSERT(tuple->len() == vectors.size());
+    for (auto i = 0u; i < vectors.size(); ++i) {
+        auto vector = vectors[i];
+        auto pos = vector->state->getSelVector()[*vectorsSelPos[i]];
+        auto data = vector->getData() + pos * vector->getNumBytesPerValue();
+        tuple->getValue(i)->copyFromColLayout(data, vector);
+    }
+}
+
+void ArrowResultCollectorLocalState::resetCursor() {
+    for (auto i = 0u; i < chunkCursors.size(); ++i) {
+        chunkCursors[i] = 0;
+    }
+}
+
+void ArrowResultCollectorSharedState::merge(std::vector<ArrowArray> localArrays) {
+    std::unique_lock lck{mutex};
+    for (auto i = 0u; i < localArrays.size(); ++i) {
+        arrays.push_back(localArrays[i]);
+    }
+}
+
+void ArrowResultCollector::executeInternal(ExecutionContext* context) {
+    auto rowBatch = std::make_unique<ArrowRowBatch>(info.columnTypes, info.chunkSize,
```

Comment:
Looks like you don't need unique_ptr here. `auto rowBatch  = ArrowRowBatch(...)` should be fine.